006-restarting-kafka-connect-connectors-and-tasks.md

```diff
@@ -0,0 +1,116 @@
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+  annotations:
+    strimzi.io/restart: "true"
+  ...
+```
+
+When applied to a `KafkaMirrorMaker2` CR, the annotation value contains the name of the connector that is restarted. For example:
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mm2-cluster
+  annotations:
+    strimzi.io/restart: "source-cluster->target-cluster.MirrorSourceConnector"
+  ...
+```
+
+### `strimzi.io/restart-task` annotation
+
+An annotation named `strimzi.io/restart-task` can be applied to the `KafkaConnector`, and `KafkaMirrorMaker2` custom resources. The value of the annotation and the CR it is applied to determines the connector task that is restarted. Once the operator successfully calls the restart REST API endpoint, the CR is patched to remove the annotation. The operator attempts to call the restart REST API endpoint once per reconcile cycle.
+
+When applied to a `KafkaConnector` CR, the annotation value contains the ID of the task that is restarted. For example:
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+  annotations:
+    strimzi.io/restart-task: "0"
+  ...
+```
+
+When applied to a `KafkaMirrorMaker2` CR, the annotation value contains the name of the connector and the ID of the task that is restarted, delimited by a `:` character. For example:
+
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mm2-cluster
+  annotations:
+    strimzi.io/restart-task: "source-cluster->target-cluster.MirrorSourceConnector:2"
+  ...
+```
+
+
+### Restarting `FAILED` connectors and tasks
+
+The `KafkaConnector` and `KafkaMirrorMaker2` operators will have new default behaviour that automatically restarts connectors or tasks that are in the `FAILED` state. For each connector or task that is in the `FAILED` state, the operator will call the Connect REST API `restart` endpoint once per reconcile cycle to attempt to restart the connector/task.
+
+To disable the new default behaviour a new annotation named `strimzi.io/auto-restart` with a value of `false` can be applied to the `KafkaConnector` and `KafkaMirrorMaker2` CRs. This will be useful when users know that external factors will cause connectors/tasks to be in the `FAILED` state and they do not want the operator attempting automatic restarts. It also means users can keep the current behaviour of the `KafkaConnector` and `KafkaMirrorMaker2` operators if they wish.
+
+For example, the following CR disables the automatic restarting of the connector and it's tasks when they are in the `FAILED` state:
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+  annotations:
+    strimzi.io/auto-restart: "false"
```

Comment:
Should this necessarily be an annotation or part of the spec instead? Do any of the maintainers have opinions on this?

Reply:
I would agree with you Sam. I think it's compared to the `KafkaConnect` resource where we have `strimzi.io/use-connector-resources: "true"` as an annotation as well ... but tbh I would say that it should be in the spec as well.

Reply:
@ajborley I have the same comment here for `KafkaMirrorMaker2` that `auto-restart` is not so clear about what is auto restarted so maybe it could be `auto-restart-connector-task` os something like that?

Reply:
I would maybe suggest `auto-restart-task` as I assume it would be clear that it would be a connector task, unless there is still some ambiguity here?

Reply:
I think the `strimzi.io/use-connector-resources` started from the idea to see how it works first before moving it into the `.spec` section. Not uncommon in Kubernetes. I think @tombentley was the main person behind it. So the question is whether to continue this way or not. 

I personally can live with both for the auto restart flag. But the manual restart should be definitely annotation.

Reply:
I totally agree about the manual restart as an annotation (it's kind of the rebalance annotation for starting the rebalance action in the `KafkaRebalance` resource) but I would like more having `auto-restart` in the spec (so `use-connector-resources` should be moved as well for having consistency for the users).

Reply:
I think the auto-restart should be part of the spec. In general we should try to limit the number and complexity of annotations. They have numerous disadvantages when compared to the spec: They're not really documented systematically, they're less discoverable and they're not validated by the apiserver. I think we will eventually move `strimzi.io/use-connector-resources` into the spec (and change the default), but likely that's with an API bump.

I also have questions about how we should communicate to the user (via the status) that the connector and/or task is being restarted. I think repeated restart is analogous to a pod crashlooping and suggests that the user needs to take action to understand/resolve a problem.

Finally, I'd have to double check, but does a connector/task transitioning to the failed state, and being restarted cause a rebalance of the tasks within connect? If so then doing that repeatedly _ad infinitum_ is not going to be a good idea, and we should have some kind of (configurable?) policy we limit the negative effects on other connectors within the same connect cluster.

Reply:
The latest commit changes the auto-restart to be part of the `KafkaConnector` and `KafkaMirrorMaker2` specs.

@tombentley These are good questions that I've been looking into today. 

I agree that we should add some status info to show the restart count, perhaps something like the following, which would increment the count every time the restart REST API is invoked and record the time when it last restarted:
```
apiVersion: strimzi.io/v1alpha1
kind: KafkaConnector
...
status:
  autoRestart: 
    count: 2
    lastRestartTimestamp: "2020-07-13T21:59:09.832Z"
```

For your second point, since Incremental Cooperative Rebalancing ([KIP-415](https://cwiki.apache.org/confluence/display/KAFKA/KIP-415%3A+Incremental+Cooperative+Rebalancing+in+Kafka+Connect)) went into Kafka 2.3, the cost of connect task rebalancing has much reduced, but I agree that we should limit repeated restarts. We could add a back-off algorithm that watches restarts and increases the number of reconcile loops between them if there have been multiple restarts in a row. What do you think?